### PR TITLE
[NTOBJSHEX] Return early from GetInfoFromPidl when pcidl is null in LPCITEMIDLIST

### PR DESCRIPTION
**Pull request by muthu90tech**
## Purpose
This fixes crash when attempting to drag n drop items from NTObject Namespace and System Registry.

_Do a quick recap of your work here._

JIRA issues: [CORE-18480](https://jira.reactos.org/browse/CORE-18480), [CORE-18481](https://jira.reactos.org/browse/CORE-18481)

## Proposed changes
Perform a null check in GetInfoFromPidel on mkid in LPCITEMIDLIST

https://user-images.githubusercontent.com/120612254/208288691-a1cd6d4d-3afa-433c-9721-c3b3d9e1d7a1.mp4

(refer video, no more crash)

### CODE REVIEW

File: dll/shellext/ntobjshex/ntobjfolder.cpp [hunk 1]
```diff
@@ -484,6 +484,12 @@ HRESULT CNtObjectFolder::GetInfoFromPidl(LPCITEMIDLIST pcidl, const NtPidlEntry
 {
     NtPidlEntry * entry = (NtPidlEntry*) &(pcidl->mkid);
```
Comment (JoachimHenze) [inline dll/shellext/ntobjshex/ntobjfolder.cpp:486]: I would strip that empty line before your "if" in both files ntobjfolder.cpp and regfolder.cpp.
Aside from that LGTM.

**Review by JoachimHenze — APPROVED** [review]

**Comment by KRosUser2** [comment]
lgtm as based on my patch.

**Review by learn-more — CHANGES_REQUESTED** [review]
This is a very unreadable way of checking `pcidl` for null, please do that instead.

**Comment by JoachimHenze** [comment]
> This is a very unreadable way of checking `pcidl` for null, please do that instead.

?
Wouldn't he need to check "pcidl->mkid" for being NULL instead?
If pcidl was indeed NULL, then his code would not work, but crash due to null-dereference already. Which is obviously not the case.

If he was to check something else than entry, then it would also make sense to move up the check.

But honestly I do not really understand your request yet, Mark.

**Comment by learn-more** [comment]
> > This is a very unreadable way of checking `pcidl` for null, please do that instead.
> 
> ? Wouldn't he need to check "pcidl->mkid" for being NULL instead? If pcidl was indeed NULL, then his code would not work, but crash due to null-dereference already. Which is obviously not the case.

obviously this is the case, since you are not dereferencing anything.

> 
> If he was to check something else than entry, then it would also make sense to move up the check.

indeed.

> 
> But honestly I do not really understand your request yet, Mark.

I am unsure what is unclear about 'xxx' is an unreadable way of doing 'yyy', please do 'yyy' instead.

If the pointer math is confusing to you, try it out with a `char*` array on something like https://godbolt.org/

**Comment by muthu90tech** [comment]
> > > This is a very unreadable way of checking `pcidl` for null, please do that instead.
> > 
> > 
> > ? Wouldn't he need to check "pcidl->mkid" for being NULL instead? If pcidl was indeed NULL, then his code would not work, but crash due to null-dereference already.
> 
> No, checking pcidl->mkid is also wrong if `pcidl` is null.
> 
The GetInfoFromPidl(pidl, &info) gets called only if pidl in not null. Please see: https://github.com/reactos/reactos/blob/master/dll/shellext/ntobjshex/regfolder.cpp#L191

So pcidl in GetInfoFromPidl can not be null.

**Comment by learn-more** [comment]
> > > > This is a very unreadable way of checking `pcidl` for null, please do that instead.
> > > 
> > > 
> > > ? Wouldn't he need to check "pcidl->mkid" for being NULL instead? If pcidl was indeed NULL, then his code would not work, but crash due to null-dereference already.
> > 
> > 
> > No, checking pcidl->mkid is also wrong if `pcidl` is null.
> 
> The GetInfoFromPidl(pidl, &info) gets called only if pidl in not null. Please see: https://github.com/reactos/reactos/blob/master/dll/shellext/ntobjshex/regfolder.cpp#L191

You have checked just one invocation: 
https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&sr=1&s=GetInfoFromPidl

> 
> So pcidl in GetInfoFromPidl can not be null.

Clearly it is being called with null, otherwise your check would not catch it either.

**Comment by muthu90tech** [comment]
> > > > > This is a very unreadable way of checking `pcidl` for null, please do that instead.
> > > > 
> > > > 
> > > > ? Wouldn't he need to check "pcidl->mkid" for being NULL instead? If pcidl was indeed NULL, then his code would not work, but crash due to null-dereference already.
> > > 
> > > 
> > > No, checking pcidl->mkid is also wrong if `pcidl` is null.
> > 
> > 
> > The GetInfoFromPidl(pidl, &info) gets called only if pidl in not null. Please see: https://github.com/reactos/reactos/blob/master/dll/shellext/ntobjshex/regfolder.cpp#L191
> 
> You have checked just one invocation: https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&sr=1&s=GetInfoFromPidl
> 
> > So pcidl in GetInfoFromPidl can not be null.
> 
> Clearly it is being called with null, otherwise your check would not catch it either.

Ok how about checking if pcidl is null before checking pcidl->mkid in GetInfoFromPidl?

Like:
```
if (!pcidl)
{
    DbgPrint("PCIDL is NULL\n");
    return E_INVALIDARG;
}
```

File: dll/shellext/ntobjshex/ntobjfolder.cpp (hunk 1 above)
Comment (HBelusca) [inline dll/shellext/ntobjshex/ntobjfolder.cpp:486]:
```suggestion
```

File: dll/shellext/ntobjshex/regfolder.cpp [hunk 2]
```diff
@@ -428,6 +428,12 @@ HRESULT CRegistryFolder::GetInfoFromPidl(LPCITEMIDLIST pcidl, const RegPidlEntry
 {
     RegPidlEntry * entry = (RegPidlEntry*) &(pcidl->mkid);
```
Comment (HBelusca) [inline dll/shellext/ntobjshex/regfolder.cpp:430]:
```suggestion
```

**Review by HBelusca — APPROVED** [review]
OK; if pcidl is NULL then wouldn't this mean the GetInfoFromPidl function is called "incorrectly"? (if it's not expected to have either of its parameters being NULL)

File: dll/shellext/ntobjshex/ntobjfolder.cpp [hunk 3]
```diff
@@ -482,7 +482,18 @@ BOOL CNtObjectFolder::IsFolder(const NtPidlEntry * info)
...
 HRESULT CNtObjectFolder::GetInfoFromPidl(LPCITEMIDLIST pcidl, const NtPidlEntry ** pentry)
 {
+    if (!pcidl)
+    {
+        DbgPrint("PCIDL is NULL\n");
+        return E_INVALIDARG;
+    }
+
     NtPidlEntry * entry = (NtPidlEntry*) &(pcidl->mkid);
+    if (!entry)
```
Comment (learn-more) [inline dll/shellext/ntobjshex/ntobjfolder.cpp:492]: This check is completely useless.

File: dll/shellext/ntobjshex/regfolder.cpp [hunk 4]
```diff
@@ -426,7 +426,18 @@ BOOL CRegistryFolder::IsFolder(const RegPidlEntry * info)
...
 HRESULT CRegistryFolder::GetInfoFromPidl(LPCITEMIDLIST pcidl, const RegPidlEntry ** pentry)
 {
+    if (!pcidl)
+    {
+        DbgPrint("PCIDL is NULL\n");
+        return E_INVALIDARG;
+    }
+
     RegPidlEntry * entry = (RegPidlEntry*) &(pcidl->mkid);
+    if (!entry)
```
[the remark this answers is not in the document]
Comment (KRosUser2) [reply dll/shellext/ntobjshex/regfolder.cpp:436]: It's indeed what's fixing the issue faced in the ticket : NULL entry, trying to access it's cb member of the next LoC

File: dll/shellext/ntobjshex/ntobjfolder.cpp (hunk 3 above)
Comment (KRosUser2) [reply dll/shellext/ntobjshex/ntobjfolder.cpp:492]: It's indeed what's fixing the issue faced in the ticket : NULL entry, trying to access it's cb member of the next LoC

File: dll/shellext/ntobjshex/ntobjfolder.cpp (hunk 1 above)
Comment (JoachimHenze) [reply dll/shellext/ntobjshex/ntobjfolder.cpp:490]: was addressed.

File: dll/shellext/ntobjshex/ntobjfolder.cpp [hunk 5]
```diff
@@ -482,8 +482,13 @@ BOOL CNtObjectFolder::IsFolder(const NtPidlEntry * info)
...
 HRESULT CNtObjectFolder::GetInfoFromPidl(LPCITEMIDLIST pcidl, const NtPidlEntry ** pentry)
 {
-    NtPidlEntry * entry = (NtPidlEntry*) &(pcidl->mkid);
+    if (!pcidl)
+    {
+        DbgPrint("PCIDL is NULL\n");
+        return E_INVALIDARG;
+    }
 
+    NtPidlEntry * entry = (NtPidlEntry*) &(pcidl->mkid);
```
Comment (learn-more) [inline dll/shellext/ntobjshex/ntobjfolder.cpp:491]: This is still a convoluted way of writing `NtPidlEntry * entry = (NtPidlEntry*)pcidl;`,
but whatever.

**Review by learn-more — APPROVED** [review]

**Comment by KRosUser2** [comment]
3 approved. Ready for merge ?

**Pull request merged**